### PR TITLE
[microTVM] Zephyr: RISCV support for Zephyr QEMU RISCV-32/64

### PR DESCRIPTION
**Pull request by mehrdadh**
This PR adds following changes:
- Add target support for testing zephyr on riscv qemu
- Add zephyr debugger for qemu
- Add minor changes left from updating zephyr 2.5 https://github.com/apache/tvm/pull/7786

Note: test_zephyr.py fully passes on qemu_riscv64 using the zephyr 2.5. However, for qemu_riscv32 you need to update zephyr to latest master branch. This issue is mentioned on zephyr: https://github.com/zephyrproject-rtos/zephyr/issues/34026

We will update this when we update our zephyr to next version.

### CODE REVIEW

**Comment by mehrdadh** [comment]
cc @areusch

File: apps/microtvm/reference-vm/base-box-tool.py [hunk 1]
```diff
@@ -358,14 +358,18 @@ def test_command(args):
 
 
 def release_command(args):
+    vm_name = f"mehrdadh/microtvm-{args.platform}"
```
Comment (areusch) [inline apps/microtvm/reference-vm/base-box-tool.py:361]: can you remove `mehrdadh/`?

File: apps/microtvm/reference-vm/base-box-tool.py [hunk 2]
```diff
@@ -358,14 +358,18 @@ def test_command(args):
...
+    if args.platform == "zephyr":
+        vm_name = f"{vm_name}-{args.zephyr_version}"
```
Comment (areusch) [inline apps/microtvm/reference-vm/base-box-tool.py:363]: can you ensure `args.zephyr_version` is not `None`, else it will release microtvm-None, I think?

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+# This file is specific to the QEMU-emulated RISCV32 microTVM board.
+
+# For TVMPlatformGenerateRandom(). Remember, these values do not need to be truly random.
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_TIMER_RANDOM_GENERATOR=y
+
+# Default 512, for operations with large floating point data. 
+CONFIG_MAIN_STACK_SIZE=2048
+
+# For floating point operations.
```
Comment (areusch) [inline apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf:27]: could you add a note to explain which error you'll see if you don't include this?

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+# specific language governing permissions and limitations
+# under the License.
+
+# This file is specific to the QEMU-emulated RISCV32 microTVM board.
+
+# For TVMPlatformGenerateRandom(). Remember, these values do not need to be truly random.
+CONFIG_TEST_RANDOM_GENERATOR=y
+CONFIG_TIMER_RANDOM_GENERATOR=y
+
+# Default 512, for operations with large floating point data.
```
Comment (areusch) [inline apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf:24]: why 512?

File: apps/microtvm/zephyr/demo_runtime/qemu-hack/qemu-system-riscv64 [hunk 5]
```diff
@@ -0,0 +1,38 @@
+#!/bin/bash -e
```
Comment (areusch) [inline apps/microtvm/zephyr/demo_runtime/qemu-hack/qemu-system-riscv64:1]: I thiiiink you can just checkin a symlink here and `$(basename $0)` will take care of this. could you try it out?

e.g. `ln -s qemu-system-i386 qemu-system-riscv64` in this dir.

File: apps/microtvm/zephyr/demo_runtime/src/main.c [hunk 6]
```diff
@@ -265,6 +265,10 @@ static uint8_t main_rx_buf[RING_BUF_SIZE_BYTES];
 // The main function of this application.
 extern void __stdout_hook_install(int (*hook)(int));
 void main(void) {
+  // TODO (mehrdadh): Update this when zephyr version was updated to 2.6. 
+  // Uncomment this for qemu_riscv32, also update zephyr to latest version.
```
Comment (areusch) [inline apps/microtvm/zephyr/demo_runtime/src/main.c:269]: Actually, can you just use the `#ifdef CONFIG_BOARD_QEMU_RISCV32`?

File: python/tvm/micro/contrib/zephyr.py [hunk 7]
```diff
@@ -571,33 +605,49 @@ def write(self, data, timeout_sec):
...
+
         self.proc = subprocess.Popen(
-            ["make", "run", f"QEMU_PIPE={self.pipe}"],
+            ["make",
+             "run",
+             f"QEMU_PIPE={self.pipe}"],
             cwd=self.base_dir,
             **self.kwargs,
         )
+        print('START DEBUG', self.debugger)
```
Comment (areusch) [inline python/tvm/micro/contrib/zephyr.py:647]: delete this line

**Review by areusch — CHANGES_REQUESTED** [review]
great work @mehrdadh ! just a couple small things to address.

cc @jossevandelm @dream-math

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf (hunk 4 above)
Comment (mehrdadh) [reply apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf:24]: based on my understanding from their doc, the default it 512. [Here](https://docs.zephyrproject.org/2.5.0/reference/kconfig/CONFIG_MAIN_STACK_SIZE.html?highlight=config_main_stack_size#cmdoption-arg-CONFIG_MAIN_STACK_SIZE) it mentioned for RISCV is 512.

File: apps/microtvm/reference-vm/base-box-tool.py (hunk 1 above)
Comment (mehrdadh) [reply apps/microtvm/reference-vm/base-box-tool.py]: Done!

File: apps/microtvm/reference-vm/base-box-tool.py (hunk 2 above)
Comment (mehrdadh) [reply apps/microtvm/reference-vm/base-box-tool.py:363]: thanks for catching this!

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf (hunk 3 above)
Comment (mehrdadh) [reply apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf]: Done!

File: apps/microtvm/zephyr/demo_runtime/qemu-hack/qemu-system-riscv64 (hunk 5 above)
Comment (mehrdadh) [reply apps/microtvm/zephyr/demo_runtime/qemu-hack/qemu-system-riscv64]: it worked!

File: apps/microtvm/zephyr/demo_runtime/src/main.c (hunk 6 above)
Comment (mehrdadh) [reply apps/microtvm/zephyr/demo_runtime/src/main.c]: that works, I think because it is an integer compare to using BOARD flag.

**Comment by tqchen** [comment]
BTW @mehrdadh here are some git tips that might be useful(try to rebase to bring the code back instead of merge) https://tvm.apache.org/docs/contribute/git_howto.html

**Comment by mehrdadh** [comment]
@tqchen Thanks for the tips! I will follow the guideline.

**Comment by areusch** [comment]
@tqchen, this method unfortunately makes the review much harder because GH loses the old commits. If you're trying to decide whether an open comment has been addressed, it's really helpful to be able to be able to see the context which generated that comment and a diff. 

I think the strategy adopted by me (and a few others) has been to rebase and force-push until a comment is placed on the PR, then do merges. If we want to consider improving the situation, I think we unfortunately may have to look into a different code review tool e.g. Gerrit (unless I misunderstand repository config settings on GH).

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf (hunk 4 above)
Comment (areusch) [reply apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf:24]: ok sounds good

File: apps/microtvm/reference-vm/base-box-tool.py (hunk 2 above)
Comment (areusch) [reply apps/microtvm/reference-vm/base-box-tool.py:363]: I think it would be better to not by default publish any version, but instead require that the user explicitly state it. otherwise, we'll wind up publishing lots of zephyr-2.5 boxes :)

also--we could probably call this platform-version, since `tlcpack/microtvm-<platform>-<platform_version>` is a pretty generic format we'll keep if we add more platforms

File: tests/micro/zephyr/test_zephyr.py [hunk 8]
```diff
@@ -59,7 +66,7 @@ def _make_sess_from_op(model, zephyr_board, west_cmd, op_name, sched, arg_bufs):
 
 
 def _make_session(model, target, zephyr_board, west_cmd, mod):
-    test_name = f"{os.path.splitext(os.path.abspath(__file__))[0]}_{model}"
+    test_name = f"{os.path.splitext(os.path.abspath(__file__))[0]}_{zephyr_board}"
```
Comment (areusch) [inline tests/micro/zephyr/test_zephyr.py:69]: why this change? seems like we should put both in the path, if anything

**Review by areusch — COMMENTED** [review]
sorry couple more small things

File: tests/micro/zephyr/test_zephyr.py (hunk 8 above)
Comment (mehrdadh) [reply tests/micro/zephyr/test_zephyr.py:69]: @areusch  so we distinguish the test based on the zephyr_board. For ex. qemu_x86, qemu_riscv32 and qemu_riscv64 all use the same `model` which is `host`, but they have different zephyr_board. zephyr_board will always be unique.

File: apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf (hunk 4 above)
Comment (mehrdadh) [reply apps/microtvm/zephyr/demo_runtime/boards/qemu_riscv32.conf:24]: Done!

File: apps/microtvm/reference-vm/base-box-tool.py (hunk 2 above)
Comment (mehrdadh) [reply apps/microtvm/reference-vm/base-box-tool.py]: I changed it to platform_version and rework the input args with sub_parser.

File: tests/micro/zephyr/test_zephyr.py (hunk 8 above)
Comment (areusch) [reply tests/micro/zephyr/test_zephyr.py:69]: oh sorry, I was misreading. this looks good

**Review by areusch — APPROVED** [review]
